### PR TITLE
snap, store: add MustBuy property to snap.Info

### PR DESCRIPTION
**Pull request by pete-woods**
Update store.Snap and store.FindSnaps to set the value of MustBuy using
recently added methods.

### CODE REVIEW

File: store/store.go [hunk 1]
```diff
@@ -428,7 +429,17 @@ func (s *SnapUbuntuStoreRepository) Snap(name, channel string, auther Authentica
 
 	s.checkStoreResponse(resp)
 
-	return infoFromRemote(searchData.Payload.Packages[0]), nil
+	pkg := searchData.Payload.Packages[0]
+
+	var purchases []*purchase
+	if auther != nil {
+		purchases, err = s.getPurchases(pkg.SnapID, channel, auther)
```
Comment (pedronis) [inline store/store.go]: do we really always need this information? I understand why it seems convenient to always ask for it but given how the store services are organized it doesn't seem a good idea

Comment (pedronis) [inline store/store.go]: and even if we must/want to do this query always, we should avoid it if we know the snap is free

**Comment by niemeyer** [comment]
From the conversation on the channel:

Gustavo Niemeyer, [04.05.16 10:55]
@pete-woods You'll probably sigh when reading this, but why do we not get the item purchases for a snap in the response of the snap request?

Pete Woods, [04.05.16 11:14]
@niemeyer At some point that will probably be useful (e.g. when we eventually support in app purchases, and the snaps can ask snappy about what the user has purchased)

Pete Woods, [04.05.16 11:14]
But right now that's a load of work that I haven't been tasked with at all

Pete Woods, [04.05.16 11:15]
So I don't feel comfortable specifying that interface purely speculatively

Pete Woods, [04.05.16 11:15]
I'd rather add that info at the time we actually use it

Pete Woods, [04.05.16 11:16]
Particularly as we maybe don't just want to clone the data structure coming from the purchase server

Pete Woods, [04.05.16 11:17]
Right now I suspect that snaps can't be given direct access to the snapd rest interface, as they could run amok

Pete Woods, [04.05.16 11:17]
But maybe there's another interface I'm not aware of

Gustavo Niemeyer, [04.05.16 11:20]
@pete-woods The question was slightly different

Pete Woods, [04.05.16 11:20]
Okay

Pete Woods, [04.05.16 11:20]
/me listens

Gustavo Niemeyer, [04.05.16 11:21]
@pete-woods Why do we need two queries for everything single find request now?

Pete Woods, [04.05.16 11:21]
@niemeyer How were you expecting it to work?  / how would you prefer it?

Pete Woods, [04.05.16 11:23]
We obviously can't get away without two queries fundamentally

Pete Woods, [04.05.16 11:23]
But I guess we could probably organise it such that they don't happen inside Snaps and FindSnaps

Gustavo Niemeyer, [04.05.16 11:25]
Why?

Gustavo Niemeyer, [04.05.16 11:26]
I mean, why can't we get away with a single query?

Pete Woods, [04.05.16 11:26]
Because the information lives on two different servers?

Gustavo Niemeyer, [04.05.16 11:27]
@pete-woods I bet it lives on three or four serves, actually :)

Pete Woods, [04.05.16 11:27]
Maybe even more than that :)

Pete Woods, [04.05.16 11:28]
But even talking in terms of logic servers, rather than physical

Pete Woods, [04.05.16 11:28]
To get all the info about a snap, detail and purchases, I need to ask two servers

Pete Woods, [04.05.16 11:31]
I admit it's weird that you're basically asked to do a database join in the client

Pete Woods, [04.05.16 11:31]
But that's the API I have to work with

Samuele Pedroni, [04.05.16 11:31]
well it's weird only assuming you need to do it all the time

Pete Woods, [04.05.16 11:32]
I guess what I'd expect is that you'd always talk to the same server

Pete Woods, [04.05.16 11:32]
But you get more details when you're authenticated

Pete Woods, [04.05.16 11:32]
But anyway, what I expect is irrelevant

Gustavo Niemeyer, [04.05.16 11:33]
@pedronis With the changes proposes in 1125 we'll do it all the time

Samuele Pedroni, [04.05.16 11:33]
yea, but it's not true

Samuele Pedroni, [04.05.16 11:34]
if the snap is free for example it doesn't make sense

Samuele Pedroni, [04.05.16 11:34]
(admittedly the server could use the same optimisation)

Samuele Pedroni, [04.05.16 11:34]
but I don't expect the server to change that way soon

Pete Woods, [04.05.16 11:34]
Okay, that's a sensible optimisation

Gustavo Niemeyer, [04.05.16 11:34]
+1

Pete Woods, [04.05.16 11:34]
Inside the Snap method

Pete Woods, [04.05.16 11:35]
But that method is basically never used afaict

Samuele Pedroni, [04.05.16 11:35]
?

Samuele Pedroni, [04.05.16 11:35]
it used when installing?

Pete Woods, [04.05.16 11:35]
Ah, okay, must have missed that

Samuele Pedroni, [04.05.16 11:35]
overlord/snapstate

Pete Woods, [04.05.16 11:35]
I can certainly make that optimisation, at any rate

Pete Woods, [04.05.16 11:36]
And a test for it

Samuele Pedroni, [04.05.16 11:36]
yes, it would be weird to fail to install something free

Samuele Pedroni, [04.05.16 11:36]
because the purchase server is down

Samuele Pedroni, [04.05.16 11:36]
(given the current structuring of the server)

Gustavo Niemeyer, [04.05.16 11:36]
Yeah

Pete Woods, [04.05.16 11:36]
Yeah, I don't expect to get the server architecture to change just for me

Gustavo Niemeyer, [04.05.16 11:37]
@pete-woods You mean, just so every user has a better experience, right? :)

Gustavo Niemeyer, [04.05.16 11:37]
@pete-woods It's okay, though.. let's work with what we have

Gustavo Niemeyer, [04.05.16 11:38]
@pete-woods Let's do the obvious optimizations, so it doesn't query when it's free, and so that if it fails it pretends there are no purchases rather than failing altogether

Gustavo Niemeyer, [04.05.16 11:38]
@pete-woods So we're still showing find results if the purchases server is off or misbehaving

Pete Woods, [04.05.16 11:38]
Okay

Gustavo Niemeyer, [04.05.16 11:38]
@pete-woods In the future we can also make the request concurrent, but let's not bother for now.. it's more complexity for exactly zero paid snaps we have atm

Pete Woods, [04.05.16 11:38]
How should I handle that, in terms of reporting the problem?

Pete Woods, [04.05.16 11:39]
Logging it doesn't seem sufficient

Samuele Pedroni, [04.05.16 11:39]
you can return both an error and the partial result, as long as daemon learns about this

Samuele Pedroni, [04.05.16 11:39]
is not an unheard of kind of style

Pete Woods, [04.05.16 11:39]
Works for me

Pete Woods, [04.05.16 11:40]
@niemeyer you on board with that, too?

Gustavo Niemeyer, [04.05.16 11:41]
@pete-woods Return the error from the purchases call, because it really failed and shouldn't pretend it works

Pete Woods, [04.05.16 11:42]
That makes sense

Gustavo Niemeyer, [04.05.16 11:42]
@pete-woods The call sites have the authority to know when to ignore and when to log though

Gustavo Niemeyer, [04.05.16 11:42]
@pete-woods Sorry, I meant when to ignore and log, and when to error out due to it

Pete Woods, [04.05.16 11:42]
But what about the Find and Snaps method

Gustavo Niemeyer, [04.05.16 11:42]
@pete-woods E.g. if you try to buy something, we should error hard if we cannot get the purchases list

Pete Woods, [04.05.16 11:42]
Of course

Gustavo Niemeyer, [04.05.16 11:42]
@pete-woods If we're returning the results of /find to the user, we should log and ignore

Samuele Pedroni, [04.05.16 11:43]
later if we find we need more fine grained detection we can add a wrapper error type to mark the case

Gustavo Niemeyer, [04.05.16 11:43]
@pedronis The partial results trick is not useful in this case, because the "partial" results of a failed purchases list would be empty anyway

Samuele Pedroni, [04.05.16 11:44]
@niemeyer I meant partial as in a list of snaps but unknown purchase status

Pete Woods, [04.05.16 11:44]
Okay, so I will do the free snap optimisation in Snap

Pete Woods, [04.05.16 11:44]
Should I do the same in FindSnaps? (But search the whole list)

Gustavo Niemeyer, [04.05.16 11:44]
@pedronis I see, yeah.. that's basically what I proposed above.. except rather than "unknown" we'll simply say "MustBuy: true"..

Gustavo Niemeyer, [04.05.16 11:44]
@pete-woods Yeah, definitely

Samuele Pedroni, [04.05.16 11:44]
@pete-woods yes, seems sensible to do it in all cases

Samuele Pedroni, [04.05.16 11:45]
the same way

Pete Woods, [04.05.16 11:45]
Cool

Samuele Pedroni, [04.05.16 11:46]
@pete-woods you could change the signature of getPurchases to take an info (I think we have already one at that point) and do the opt inside it

Samuele Pedroni, [04.05.16 11:47]
or something along those lines

Gustavo Niemeyer, [04.05.16 11:48]
Sounds reasonable

Samuele Pedroni, [04.05.16 11:49]
may by tricky in terms of marking free as result there

Samuele Pedroni, [04.05.16 11:49]
but there should be some way

Gustavo Niemeyer, [04.05.16 11:50]
@pedronis Free is already handled by len(Prices) == 0

Samuele Pedroni, [04.05.16 11:50]
ah, yes

Samuele Pedroni, [04.05.16 11:50]
so empty would work for both cases

Samuele Pedroni, [04.05.16 11:51]
anyway in FindSnaps the question is if they are all free

Gustavo Niemeyer, [04.05.16 11:52]
@pedronis Prices are actually returned by the first query, I think

Samuele Pedroni, [04.05.16 11:52]
yes

Samuele Pedroni, [04.05.16 11:53]
mostly wondering about code org

Gustavo Niemeyer, [04.05.16 11:53]
@pedronis So we won't say "free" even if we cannot talk to the purchases server

Gustavo Niemeyer, [04.05.16 11:53]
Which is a good thing

Samuele Pedroni, [04.05.16 11:53]
but @pete_woods can see what works

Gustavo Niemeyer, [04.05.16 11:54]
@pedronis @pete_woods Saying "Price = 1BRL" if it was already bought is a pretty minor wart for an offline server

Samuele Pedroni, [04.05.16 11:55]
yea, mostly thinking about the asymmetry of getPurchases vs getAllPurchases (we don't have a bulk query for not all them afaiu)

Pete Woods, [04.05.16 11:57]
@pedronis that is correct

Pete Woods, [04.05.16 11:57]
It would be much more useful to be able to request purchases for a specific set of snap IDs

Gustavo Niemeyer, [04.05.16 11:58]
@pete_woods Perhaps invert the API of the store, at least

Gustavo Niemeyer, [04.05.16 11:58]
@pete_woods @pedronis decoratePurchases(snaps)

Samuele Pedroni, [04.05.16 11:58]
that would work

Samuele Pedroni, [04.05.16 11:59]
(and fit things better)

Pete Woods, [04.05.16 11:59]
Works for me

Gustavo Niemeyer, [04.05.16 11:59]
@pete_woods It can optimize for len(snaps) == 1 internally

**Comment by niemeyer** [comment]
Closing this one per discussion above.

**Pull request closed without merging**